File: src/main/java/bad/robot/radiate/teamcity/TeamCity.java
package bad.robot.radiate.teamcity;

import bad.robot.http.Headers;
import bad.robot.http.HttpClient;
import bad.robot.http.HttpResponse;
import bad.robot.radiate.Unmarshaller;

import java.net.URL;

import static bad.robot.http.HeaderList.headers;
import static bad.robot.http.HeaderPair.header;
import static bad.robot.radiate.teamcity.TeamCityEndpoint.projectsEndpoint;

class TeamCity {

    private final Headers headers = headers(header("Accept", "application/json"));
    private final Server server;
    private final HttpClient http;
    private final Unmarshaller<HttpResponse, Iterable<Project>> unmarshaller;

    public TeamCity(Server server, HttpClient http, Unmarshaller<HttpResponse, Iterable<Project>> unmarshaller) {
        this.server = server;
        this.http = http;
        this.unmarshaller = unmarshaller;
    }

    public Iterable<Project> retrieveProjects() {
        URL url = server.urlFor(projectsEndpoint);
        HttpResponse response = http.get(url, headers);
        if (response.ok())
            return unmarshaller.unmarshall(response);
        throw new UnexpectedResponse(url, response);
    }

    public void retrieveFullProjectInformation(Iterable<Project> projects) {
        for (Project project : projects) {
            URL url = server.urlFor(project);
            HttpResponse response = http.get(url, headers);
            System.out.println(response.getContent().asString() +"\n\n");
        }
    }

}

File: src/main/scala/bad/robot/radiate/teamcity/TeamCity.scala
package bad.robot.radiate.teamcity

import bad.robot.http.{Headers, HttpClient, HttpResponse}
import bad.robot.radiate.Unmarshaller
import bad.robot.http.HeaderList._
import bad.robot.http.HeaderPair.header
import bad.robot.radiate.teamcity.BuildLocatorBuilder.latest
import bad.robot.radiate.teamcity.BuildLocatorBuilder.running
import bad.robot.radiate.teamcity.TeamCityEndpoint.projectsEndpoint

import scala.collection.JavaConversions._

case class TeamCity(server: Server, http: HttpClient, projects: Unmarshaller[HttpResponse, java.lang.Iterable[Project]], project: Unmarshaller[HttpResponse, Project], build: Unmarshaller[HttpResponse, Build]) {

  private final val accept: Headers = headers(header("Accept", "application/json"))

  def retrieveProjects: java.lang.Iterable[Project] = {
    val url = server.urlFor(projectsEndpoint)
    val response = http.get(url, accept)
    if (response.ok)
      return projects.unmarshall(response)
    throw new UnexpectedResponse(url, response)
  }

  def retrieveBuildTypes(projects: java.lang.Iterable[Project]): java.lang.Iterable[BuildType] = {
    val expanded = projects.map(
      (project) => {
        val url = server.urlFor(project)
        val response = http.get(url, accept)
        if (response.ok)
          Right(TeamCity.this.project.unmarshall(response))
        else
          Left(new UnexpectedResponse(url, response))
      })

    val exceptions = expanded.collect{ case Left(exception) => exception }
    if (!exceptions.isEmpty)
      throw exceptions.head

    asJavaIterable(expanded.collect{ case Right(p) => p }.flatMap{ p => p })
  }

  def retrieveLatestBuild(buildType: BuildType): Build = {
    val url = server.urlFor(running(buildType))
    val response = http.get(url, accept)
    if (response.ok)
      return build.unmarshall(response)
    if (response.getStatusCode == 404)
      return retrieveBuild(latest(buildType))
    throw new UnexpectedResponse(url, response)
  }

  private def retrieveBuild(locator: BuildLocatorBuilder): Build = {
    val url = server.urlFor(locator)
    val response = http.get(url, accept)
    if (response.ok)
      return build.unmarshall(response)
    throw new UnexpectedResponse(url, response)
  }

}

